processing/src/main/java/org/apache/druid/query/filter/ArrayContainsFilter.java

```diff
@@ -0,0 +1,677 @@
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.TypeSignature;
+import org.apache.druid.segment.column.TypeStrategy;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.segment.filter.PredicateValueMatcherFactory;
+import org.apache.druid.segment.filter.ValueMatchers;
+import org.apache.druid.segment.index.AllUnknownBitmapColumnIndex;
+import org.apache.druid.segment.index.BitmapColumnIndex;
+import org.apache.druid.segment.index.semantic.ArrayElementIndexes;
+import org.apache.druid.segment.index.semantic.StringValueSetIndexes;
+import org.apache.druid.segment.index.semantic.ValueIndexes;
+import org.apache.druid.segment.nested.StructuredData;
+import org.apache.druid.segment.vector.VectorColumnSelectorFactory;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ArrayContainsFilter extends AbstractOptimizableDimFilter implements Filter
+{
+  private final String column;
+  private final ColumnType elementMatchValueType;
+  private final Object elementMatchValue;
+  private final ExprEval<?> elementMatchValueEval;
+
+  @Nullable
+  private final FilterTuning filterTuning;
+  private final DruidPredicateFactory predicateFactory;
+
+  @JsonCreator
+  public ArrayContainsFilter(
+      @JsonProperty("column") String column,
+      @JsonProperty("elementMatchValueType") ColumnType elementMatchValueType,
+      @JsonProperty("elementMatchValue") Object elementMatchValue,
+      @JsonProperty("filterTuning") @Nullable FilterTuning filterTuning
+  )
+  {
+    if (column == null) {
+      throw InvalidInput.exception("Invalid array_contains filter, column cannot be null");
+    }
+    this.column = column;
+    if (elementMatchValueType == null) {
+      throw InvalidInput.exception("Invalid array_contains filter on column [%s], elementMatchValueType cannot be null", column);
+    }
+    this.elementMatchValueType = elementMatchValueType;
+    this.elementMatchValue = elementMatchValue;
+    this.elementMatchValueEval = ExprEval.ofType(ExpressionType.fromColumnTypeStrict(elementMatchValueType), elementMatchValue);
+    this.filterTuning = filterTuning;
+    this.predicateFactory = new ArrayContainsPredicateFactory(elementMatchValueEval);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    final TypeStrategy<Object> typeStrategy = elementMatchValueEval.type().getStrategy();
+    final int size = typeStrategy.estimateSizeBytes(elementMatchValueEval.value());
+    final ByteBuffer valueBuffer = ByteBuffer.allocate(size);
+    typeStrategy.write(valueBuffer, elementMatchValueEval.value(), size);
+    return new CacheKeyBuilder(DimFilterUtils.ARRAY_CONTAINS_CACHE_ID)
+        .appendByte(DimFilterUtils.STRING_SEPARATOR)
```

Comment:
Shouldn't `CacheKeyBuilder#appendString` mean we don't need `STRING_SEPARATOR`?

Although, looking at `CacheKeyBuilder#appendString`, I think it needs to itself include either `STRING_SEPARATOR` or a length, in case a string contains a byte equal to a valid type code.

I guess this doesn't have to be done in this patch, but we should fix it.